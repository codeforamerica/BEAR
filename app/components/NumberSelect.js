// @flow
import React, { Component } from 'react';

type Props = {
  labelText: string,
  group: string,
  minNumber: number,
  maxNumber: number,
  onNumberSelect: (string, number) => void,
  selectedNumber: number
};

export default class NumberSelect extends Component<Props> {
  handleNumberSelect = (e: SyntheticEvent<HTMLSelectElement>) => {
    const { group, onNumberSelect } = this.props;
    const selectedValue = parseInt(e.currentTarget.selectedOptions[0].text, 10);
    onNumberSelect(group, selectedValue);
  };

  numberRange = (numberThreshold: number, maximumNumber: number) => {
    const list = [];
    // eslint-disable-next-line no-plusplus
    for (let i = numberThreshold; i <= maximumNumber; i++) {
      list.push(i);
    }
    return list;
  };

  render() {
    const {
      labelText,
      group,
      minNumber,
      maxNumber,
      selectedNumber
    } = this.props;
    return (
      <div>
        {/* Below keeps failing linter despite having associated control (and passing screenreader) */}
        {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
        {/* eslint-disable-next-line jsx-a11y/label-has-for */}
        <label htmlFor="number-select" className="sr-only">
          {labelText}
        </label>
        <div className="select">
          <select
            className="select__element"
            onChange={this.handleNumberSelect}
            id={`${group}-select`}
            value={selectedNumber}
          >
            {this.numberRange(minNumber, maxNumber).map(num => (
              <option key={num} value={num}>
                {num}
              </option>
            ))}
          </select>
        </div>
      </div>
    );
  }
}
