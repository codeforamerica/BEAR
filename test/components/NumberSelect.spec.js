import sinon from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import NumberSelect from '../../app/components/NumberSelect';

Enzyme.configure({ adapter: new Adapter() });
const sandbox = sinon.createSandbox();

function setup() {
  const onNumberSelectSpy = sandbox.spy();
  const component = shallow(
    <NumberSelect
      onNumberSelect={onNumberSelectSpy}
      group="age"
      minNumber={1}
      maxNumber={3}
    />
  );
  return {
    component,
    onNumberSelectSpy
  };
}

afterEach(() => {
  sandbox.restore();
});

describe('NumberSelect component', () => {
  it('selecting a number should call onNumberSelect with numeric value', () => {
    const { component, onNumberSelectSpy } = setup();
    const ageSelect = component.find('#age-select');
    const fakeEvent = {
      currentTarget: {
        value: '3',
        selectedOptions: [{ text: '3' }]
      }
    };
    ageSelect.simulate('change', fakeEvent);
    expect(onNumberSelectSpy.called).toBe(true);
    expect(onNumberSelectSpy.args[0][1]).toEqual(3);
  });

  it('should populate dropdown with numbers between max and min', () => {
    const { component } = setup();
    const ageSelect = component.find('#age-select');
    const ageOptions = ageSelect.find('option').map(opt => {
      return opt.props().value;
    });
    expect(ageOptions).toEqual([1, 2, 3]);
  });

  it('should match exact snapshot', () => {
    const { component } = setup();
    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
