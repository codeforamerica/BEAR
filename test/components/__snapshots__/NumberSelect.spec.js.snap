// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberSelect component should match exact snapshot 1`] = `
<div>
  <label
    className="sr-only"
    htmlFor="number-select"
  />
  <div
    className="select"
  >
    <select
      className="select__element"
      id="age-select"
      onChange={[Function]}
    >
      <option
        value={1}
      >
        1
      </option>
      <option
        value={2}
      >
        2
      </option>
      <option
        value={3}
      >
        3
      </option>
    </select>
  </div>
</div>
`;
